game/groundforces/template_loader.py

```diff
@@ -0,0 +1,361 @@
+
+
+# Custom Dumper to fix pyyaml indent https://github.com/yaml/pyyaml/issues/234
+class MappingDumper(yaml.Dumper):
+    def increase_indent(self, flow: bool = False, *args: Any, **kwargs: Any) -> None:
+        return super().increase_indent(flow=flow, indentless=False)
+
+
+TEMPLATE_TYPES = {
+    GroupRole.AntiAir: AntiAirTemplate,
+    GroupRole.Building: BuildingTemplate,
+    GroupRole.Naval: NavalTemplate,
+    GroupRole.GroundForce: GroundForceTemplate,
+    GroupRole.Defenses: DefensesTemplate,
+}
+
+
+class TemplateLoader:
+    # list of templates per category. e.g. AA or similar
+    _templates: dict[str, GroundObjectTemplate] = {}
+
+    def __init__(self) -> None:
+        self._templates = {}
+
+    def initialize(self) -> None:
+        if not self._templates:
+            with logged_duration("Loading templates"):
+                self.load_templates()
+
+    @property
+    def templates(self) -> Iterator[GroundObjectTemplate]:
+        self.initialize()
+        yield from self._templates.values()
+
+    def load_templates(self) -> None:
+        """This will load all pre-loaded templates from a pickle file.
+        If pickle can not be loaded it will import and dump the templates"""
+        # We use a pickle for performance reasons. Importing takes many seconds
+        file = Path(persistency.base_path()) / TEMPLATE_DUMP
+        if file.is_file():
+            # Load from pickle if existing
+            with file.open("rb") as f:
+                try:
+                    version, self._templates = pickle.load(f)
+                    # Check if the game version of the dump is identical to the current
+                    if version == VERSION:
+                        return
+                except Exception as e:
+                    logging.error(f"Error {e} reading templates dump. Recreating.")
+        # If no dump is available or game version is different create a new dump
+        self.import_templates()
+
+    def import_templates(self) -> None:
+        """This will import all templates from the template folder
+        and dumps them to a pickle"""
+        mappings: dict[str, list[TemplateMapping]] = {}
+        with logged_duration("Parsing mapping yamls"):
+            for file in Path(TEMPLATE_DIR).rglob("*.yaml"):
+                if not file.is_file():
+                    continue
+                with file.open("r", encoding="utf-8") as f:
+                    mapping_dict = yaml.safe_load(f)
+
+                template_map = TemplateMapping.from_dict(mapping_dict, f.name)
+
+                if template_map.template_file in mappings:
+                    mappings[template_map.template_file].append(template_map)
+                else:
+                    mappings[template_map.template_file] = [template_map]
+
+        with logged_duration(f"Loading templates multithreaded"):
+            with ThreadPoolExecutor(max_workers=4) as exe:
```

Comment:
No need to limit this to 4. It'll automatically limit to the number of threads the machine has available. No sense leaving 12 of my cores waiting for these 4 to finish :)